### PR TITLE
[POC] Add APK package building capabilities to OpenWrt

### PR DESCRIPTION
**Pull request by aparcar**
APK is the main package manager of Alpine Linux. This is just a proof of concept to double create package for OPKG and APK in parrallel.

(Personal) motivation is a long time replacement of OPKG with APK and join forces with Alpine Linux to create a great package manager for embedded devices rather than maintaining OPKG.

The scripts is heavily insprired by Alpines abuild script to create repositores.

### CODE REVIEW

File: package/utils/abuild-tar/Makefile [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+HOST_BUILD_DEPENDS:=libressl/host zlib/host
+
+include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/host-build.mk
+
+define Package/abuild-tar
+  SECTION:=base
+  CATEGORY:=Base system
+  TITLE:=abuild-tar util
+  DEPENDS:=+zlib +libopessl +libthread
```
Comment (doverride) [inline package/utils/abuild-tar/Makefile:26]: should that be +libope**n**ssl? and +lib**p**thread?

Comment (aparcar) [reply package/utils/abuild-tar/Makefile]: Thanks, didn't device test this. The package is really just a build requirement and likely never device installed. It's not part of tools/ as it requires the SSL lib.

**Comment by champtar** [comment]
Do you have a feature comparison between OPKG and APK ? (just curious)

**Comment by neheb** [comment]
As far as I can see, openssl is only used for SHA1 hashes. Seems wasteful too. They have ENGINE support for some reason.

File: package/utils/abuild-tar/Makefile [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+HOST_LDFLAGS += -lpthread
+
+define Host/Compile
+	$(MAKE) -C $(HOST_BUILD_DIR) $(HOST_MAKE_FLAGS) abuild-tar abuild-sign functions.sh
+endef
+
+define Package/abuild-tar/install
+	$(INSTALL_DIR) $(1)/usr/bin
+	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/bin/abuild-tar $(1)/bin/abuild-tar
+
```
Comment (adschm) [inline package/utils/abuild-tar/Makefile:39]: Unneeded empty line.

File: package/utils/abuild-tar/Makefile [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+endef
+
+define Host/Install
+	$(INSTALL_BIN) $(HOST_BUILD_DIR)/abuild-tar $(STAGING_DIR_HOST)/bin/abuild-tar
+endef
+
+$(eval $(call BuildPackage,abuild-tar))
+$(eval $(call HostBuild))
```
Comment (adschm) [inline package/utils/abuild-tar/Makefile:47]: Is this correct without an argument?

Comment (neheb) [reply package/utils/abuild-tar/Makefile:47]: Yes

File: package/utils/abuild-tar/Makefile (hunk 2 above)
Comment (aparcar) [reply package/utils/abuild-tar/Makefile:39]: Thanks!

**Comment by aparcar** [comment]
> As far as I can see, openssl is only used for SHA1 hashes. Seems wasteful too. They have ENGINE support for some reason.

I'm sure upstream is happy if you provide a patch :)

There are already patches to replace it with SHA256 but still using openssl

**Comment by aparcar** [comment]
I just updated to the latest development version of APK which introduces some major changes:

* abuild-tar is no longer required since apk can now pack it's own packages
* The signing keys are no longer RSA based but use EC instead
* An entirely new database for package indexing is used which is optimized for embedded devices

This won't fly at this point, however things are happening.

**Comment by aparcar** [comment]
The latest push includes changes which allow building APK as a host and therefore actually create packages. The `apk` patch will shortly be send to upstream.

**Comment by aparcar** [comment]
Odd, I didn't actively close this, must have been some GitHub magic.

**Pull request closed without merging**